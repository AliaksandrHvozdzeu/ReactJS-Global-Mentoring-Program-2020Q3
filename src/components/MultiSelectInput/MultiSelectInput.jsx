import React, { useState } from 'react';
import PropTypes from 'prop-types';
import MultiSelect from 'react-multi-select-component';
import Constants from '../../constants';
import './style.css';

const buildValues = (elem) => elem.map((e) => ({ label: e, value: e }));

const ArrowRenderer = ({ expanded }) => (expanded ? <img className="arrow" src={Constants.ARROWS[0]} alt="arrow_up"/> :
  <img className="arrow" src={Constants.ARROWS[1]} alt="arrow_down"/>);

const MultiSelectInput = ({ title, options, onAction, preselected }) => {

  const [selected, setSelected] = useState(buildValues(preselected) || []);

  const onChange = (selectedElements) => {
    onAction(selectedElements);
    setSelected(selectedElements);
  };

  return (
    <label htmlFor="genres" className="genres" id="malty">
      {title}
      <MultiSelect
        id="genres"
        ArrowRenderer={ArrowRenderer}
        options={options}
        value={selected}
        onChange={onChange}
        hasSelectAll={false}
        labelledBy="malty"/>
    </label>
  );
};

ArrowRenderer.propTypes = {
  expanded: PropTypes.func.isRequired
};

MultiSelectInput.propTypes = {
  title: PropTypes.string.isRequired,
  onAction: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired
    }),
  ).isRequired,
  preselected: PropTypes.arrayOf(PropTypes.string)
};

MultiSelectInput.defaultProps = {
  preselected: []
};


export default MultiSelectInput;
