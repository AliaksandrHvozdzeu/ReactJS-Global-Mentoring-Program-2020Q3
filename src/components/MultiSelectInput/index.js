import MultiSelectInput from './js/MultiSelectInput';

export default MultiSelectInput;
