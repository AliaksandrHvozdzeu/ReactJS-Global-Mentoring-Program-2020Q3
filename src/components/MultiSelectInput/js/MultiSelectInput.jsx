import React from 'react';
import PropTypes from 'prop-types';
import MultiSelect from 'react-multi-select-component';
import Constants from '../../../constants';
import '../css/MultiSelectInput.css';
import ErrorMessage from '../../ErrorMessage';

const ArrowRenderer = ({ expanded }) => (expanded ?
  <img className="arrow" src={Constants.ARROWS[0]} alt="arrow_up"/> :
  <img className="arrow" src={Constants.ARROWS[1]} alt="arrow_down"/>);

const buildValues = (elems) => elems.map((e) => ({ label: e, value: e }));

export default function MultiSelectInput({ title, options, preselected, onAction, onBlur, error }) {

  const selected = buildValues(preselected) || [];

  return (
    <>
      <label htmlFor="genres" className="genres" id="malty">
        {title}
        <MultiSelect
          id="genres"
          ArrowRenderer={ArrowRenderer}
          options={options}
          value={selected}
          onChange={onAction}
          hasSelectAll={false}
          labelledBy="malty"
          onBlur={onBlur}/>
      </label>
      <ErrorMessage error={error}/>
    </>
  );

}

ArrowRenderer.propTypes = {
  expanded: PropTypes.bool.isRequired,
  error: PropTypes.string,
  onBlur: PropTypes.func,
};

MultiSelectInput.propTypes = {
  title: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
    }),
  ).isRequired,
  preselected: PropTypes.arrayOf(PropTypes.string),
  onAction: PropTypes.func.isRequired,
};

MultiSelectInput.defaultProps = {
  preselected: [],
};
