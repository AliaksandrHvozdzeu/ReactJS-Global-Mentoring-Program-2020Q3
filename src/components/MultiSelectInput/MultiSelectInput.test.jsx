import React from 'react';
import rerender from 'react-test-renderer';
import {render, fireEvent, screen} from '@testing-library/react';
import Constants from '../../constants';
import MultiSelectInput from './js/MultiSelectInput';

describe('when Multi Select Input', () => {

  const title = 'title';
  const options = Constants.DEFAULT;
  const preselected = ['Comedy', 'Horror', 'Documentary'];
  const onAction = () => {
  };
  const onBlur = () => {
  };
  const error = 'error';

  test('then snapshot created', () => {
    const component = rerender.create(<MultiSelectInput title={title}
                                                        options={options}
                                                        preselected={preselected}
                                                        onAction={onAction}
                                                        onBlur={onBlur}
                                                        error={error}/>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Preselected is empty', () => {
    const emptyPreselected = {};
    expect(emptyPreselected).toEqual({});
  });

  test('should call handler on change', () => {
    render(<MultiSelectInput title={title}
                             options={options}
                             preselected={preselected}
                             onAction={onAction}
                             onBlur={onBlur}
                             error={error}/>);
    const value = screen.getByText('Comedy, Horror, Documentary');
    expect(value).toEqual('Comedy, Horror, Documentary');
  });

  test('on change arrow image',() => {
    render(<MultiSelectInput title={title}
                             options={options}
                             preselected={preselected}
                             onAction={onAction}
                             onBlur={onBlur}
                             error={error}/>);
    const arrow = screen.getByAltText('arrow_down');
    fireEvent.click(arrow);
    expect(screen.getByAltText('arrow_up')).toBeTruthy();
  });

  test('on empty value',() => {
    render(<MultiSelectInput title={title}
                             options={options}
                             value={[]}
                             preselected={[]}
                             onAction={onAction}
                             onBlur={onBlur}
                             error={error}/>);
    const value = screen.getByText('Select...');
    expect(value).toContain('<span class="gray">Select...</span>');
  });

});
