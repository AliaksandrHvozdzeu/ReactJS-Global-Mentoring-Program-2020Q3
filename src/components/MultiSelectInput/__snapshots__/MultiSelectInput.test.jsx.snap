// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when Multi Select Input then snapshot created 1`] = `
Array [
  <label
    className="genres"
    htmlFor="genres"
    id="malty"
  >
    title
    <div
      className="go2139093995 multi-select"
    >
      <div
        aria-expanded={false}
        aria-labelledby="malty"
        aria-readonly={true}
        className="go2646822163 dropdown-container"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        tabIndex={0}
      >
        <div
          className="go2642161244 dropdown-heading"
          onClick={[Function]}
        >
          <div
            className="dropdown-heading-value"
          >
            <span>
              Comedy, Horror, Documentary
            </span>
          </div>
          <img
            alt="arrow_down"
            className="arrow"
            src="/images/arrow_down.png"
          />
        </div>
      </div>
    </div>
  </label>,
  <p
    className="input-error"
  >
    error
  </p>,
]
`;
